### PR TITLE
Callum/van 488 text to image results UI

### PR DESCRIPTION
**Pull request by callumflack**

### CODE REVIEW

**Comment by dmitryvlass** [comment]
@callumflack , could you reduce this space for mobile?

image

**Comment by dmitryvlass** [comment]
@callumflack , we were talking yesterday about the blinking images. It happens because we create multiple new user objects [here](https://github.com/corsali/vana-portrait-demo/blob/callum/van-488-text-to-image-results-ui/pages/index.js#L157)

So the user lifecycle looks next:
--- index.js renders 
1. Set the initial user
https://github.com/corsali/vana-portrait-demo/blob/callum/van-488-text-to-image-results-ui/pages/index.js#L21
2. Fetch the initial exhibits and rewrite the user object
--- refreshExhibits()
3. Create [the new user object](https://github.com/corsali/vana-portrait-demo/blob/callum/van-488-text-to-image-results-ui/pages/index.js#L124)

Because we update the reference to the user object, all dependent components also refresh, causing re-rendering of the whole DOM tree. It can be solved by using the primitives for the useState, updating only what has to be updated and passing only required data to the nested components.

File: components/prompt/Generator.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  };
+
+  useEffect(() => {
+    if (prompt.length > 20) {
+      setValidPrompt(meRegex.test(prompt))
+    }
+    if (isLoading) {
+      setValidPrompt(meRegex.test(prompt))
+    }
+  }, [prompt])
```
Comment (dmitryvlass) [inline components/prompt/Generator.js:44]: You don't change `setValidPrompt` back when the prompt is valid. Could you handle the validation inside the <textarea onChange={handlePromptChange} ... /> ?

File: components/prompt/IdeasMessage.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { PromptIdeas } from "components";
+
+const IdeasMessage = () => {
+  return (
+    <div className="space-y-4">
+      <h2 className="">Stuck for ideas?</h2>
```
Comment (dmitryvlass) [inline components/prompt/IdeasMessage.js:6]:
```suggestion
      <h2>Stuck for ideas?</h2>
```

File: components/prompt/IdeasMessage.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      <p className="">Sometimes, all it takes is attention.</p>
```
Comment (dmitryvlass) [inline components/prompt/IdeasMessage.js:7]:
```suggestion
      <p>Sometimes, all it takes is attention.</p>
```

File: components/prompt/IdeasMessage.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+      <PromptIdeas className="pl-3" />
+      <p className="">
```
Comment (dmitryvlass) [inline components/prompt/IdeasMessage.js:9]:
```suggestion
      <p>
```

File: components/prompt/IdeasMessage.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+        Still stuck? <a href="mailto:neil@vana.com" style={{ textDecoration: "underline"}}>Ask Neil.</a>{" "}
```
Comment (dmitryvlass) [inline components/prompt/IdeasMessage.js:10]:
```suggestion
        Still stuck? <a href="mailto:neil@vana.com" style={{ textDecoration: "underline"}}>Ask Neil.</a>
```

File: components/prompt/PromptAwaitingMessage.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { PromptIdeas } from "components";
+
+const PromptAwaitingMessage = () => {
+  return (
+    <div className="space-y-4">
+      <h2 className="">Takes about a minute</h2>
```
Comment (dmitryvlass) [inline components/prompt/PromptAwaitingMessage.js:6]:
```suggestion
      <h2>Takes about a minute</h2>
```

File: components/prompt/PromptAwaitingMessage.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      <p className="">While you're waiting, here's some other ideas:</p>
```
Comment (dmitryvlass) [inline components/prompt/PromptAwaitingMessage.js:7]:
```suggestion
      <p>While you're waiting, here's some other ideas:</p>
```

File: components/prompt/PromptAwaitingMessage.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      <PromptIdeas className="pl-3" />
+      <p className="">
```
Comment (dmitryvlass) [inline components/prompt/PromptAwaitingMessage.js:9]:
```suggestion
      <p>
```

**Review by dmitryvlass — CHANGES_REQUESTED** [review]

File: pages/index.js [hunk 9]
```diff
@@ -74,93 +81,52 @@ export default function Home() {
...
-      setUser(newUser);
+  // Get Text to Image exhibit images
+  const populateTextToImageExhibits = useCallback(async (token) => {
+    async function refreshImages() {
+      const images = await getTextToImageUserExhibits(token);
 
-      Object.keys(newUser.exhibits).length &&
-        setRandomExhibitImages(getRandomImages(3, newUser.exhibits));
-    };
+      setTextToImageExhibitImages(images);
```
Comment (dmitryvlass) [inline pages/index.js:96]: We need to reverse the images to show the last generated images on the top.

```suggestion
      setTextToImageExhibitImages(images.reverse());
```

**Comment by callumflack** [comment]
@tnunamak @dmitryvlass heads up, I misunderstood how the Generator's `vanaPost` worked. So still have this todo:

- [ ] showing awaiting UI while `vanaPost` is running

**Comment by callumflack** [comment]
Simple UI state when a user has not yet created any `textToImageExhibitImages`:

image

File: components/prompt/Generator.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState(null);
+  const [validPrompt, setValidPrompt] = useState(true);
+  const [showIdeas, setShowIdeas] = useState(false);
+
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+
+    // isLoading triggers the awaiting dialog. We want to delay it a little so that if there's an error, the dialog doesn't flash up
+    setTimeout(() => { setIsLoading(true) }, 1000);
```
[the remark this answers is not in the document]
Comment (callumflack) [reply components/prompt/Generator.js:20]: @dmitryvlass based on chats about this, I misunderstood how `vanaPost('jobs/text-to-image')` worked. So we now know there's no reason to show a dialog at this point (the timeout above is related to this). So this is now deleted.

**Review by dmitryvlass — APPROVED** [review]

**Pull request merged**